File: backend/src/modules/CompanyProfile/CompanyProfile.Domain/ValueObjects/GeoCoordinates.ts
import { InvalidGeoCoordinatesException } from '../Exceptions/InvalidGeoCoordinatesException';

export class GeoCoordinates {
  public constructor(public readonly latitude: number, public readonly longitude: number) {
    if (!Number.isFinite(latitude) || !Number.isFinite(longitude)) {
      throw new InvalidGeoCoordinatesException('Las coordenadas deben ser valores numéricos finitos.');
    }
    if (latitude < -90 || latitude > 90) {
      throw new InvalidGeoCoordinatesException('La latitud debe estar entre -90 y 90.');
    }
    if (longitude < -180 || longitude > 180) {
      throw new InvalidGeoCoordinatesException('La longitud debe estar entre -180 y 180.');
    }
  }

  public equals(other: unknown): boolean {
    return other instanceof GeoCoordinates && other.latitude === this.latitude && other.longitude === this.longitude;
  }
}

File: backend/modules/CompanyProfile/CromaticaCreativa.Modules.CompanyProfile.Domain/ValueObjects/GeoCoordinates.cs
using CromaticaCreativa.Modules.CompanyProfile.Domain.Exceptions;

namespace CromaticaCreativa.Modules.CompanyProfile.Domain.ValueObjects;

public readonly record struct GeoCoordinates
{
    public GeoCoordinates(double latitude, double longitude)
    {
        if (double.IsNaN(latitude) ||
            double.IsNaN(longitude) ||
            latitude is < -90 or > 90 ||
            longitude is < -180 or > 180)
        {
            throw new InvalidGeoCoordinatesException(latitude, longitude);
        }

        Latitude = latitude;
        Longitude = longitude;
    }

    public double Latitude { get; }

    public double Longitude { get; }
}
